code/modules/reagents/machinery/pump.dm

```diff
@@ -0,0 +1,188 @@
+		var/image/I = image(icon, "[icon_state]-volume")
+		I.color = reagents.get_color()
+		add_overlay(I)
+	add_overlay("[icon_state]-glass")
+
+	if(open)
+		add_overlay("[icon_state]-open")
+		if(istype(cell))
+			add_overlay("[icon_state]-cell")
+
+	icon_state = "[initial(icon_state)][on ? "-running" : ""]"
+
+/obj/machinery/pump/process()
+	if(!on)
+		return
+	
+	if(!anchored || !(cell?.use(active_power_usage)))
+		toggle(FALSE)
+		return
+
+	pump_reagents()
+	update_icon()
+
+	if(Output.get_pairing())
+		reagents.trans_to_holder(Output.reagents, Output.reagents.maximum_volume)
+		if(prob(5))
+			visible_message("<span class='notice'>\The [src] gurgles as it exports fluid.</span>")
+
+
+// Toggles the power state, if possible.
+// Returns TRUE/FALSE on power state changing
+// var/target = target power state
+// var/message = TRUE/FALSE whether to make a message about state change
+/obj/machinery/pump/proc/toggle(var/target, var/message = TRUE)
+	if(target == on)
+		return FALSE
+
+	if(!on && (!(cell?.check_charge(active_power_usage)) || !anchored))
+		return FALSE
+
+	on = !on
+	update_icon()
+	if(message)
+		if(on)
+			message = SPAN_NOTICE("\The [src] turns on.")
+		else
+			message = SPAN_NOTICE("\The [src] shuts down.")
+		visible_message(message)
+	return TRUE
+
+/obj/machinery/pump/attack_robot(mob/user)
+	return attack_hand(user)
+
+/obj/machinery/pump/attack_ai(mob/user)
+	if(!toggle(!on))
+		to_chat(user, "<span class='notice'>You try to toggle \the [src] but it does not respond.</span>")
+
+/obj/machinery/pump/attack_hand(mob/user)
+	if(open && istype(cell))
+		user.put_in_hands(cell)
+		cell.add_fingerprint(user)
+		cell.update_icon()
+		cell = null
+		toggle(FALSE)
+		to_chat(user, "<span class='notice'>You remove the power cell.</span>")
+		return
+
+	if(!toggle(!on))
+		to_chat(user, "<span class='notice'>You try to toggle \the [src] but it does not respond.</span>")
+
+/obj/machinery/pump/attackby(obj/item/weapon/W, mob/user)
+	. = TRUE
+	if(W.is_screwdriver() && !open)
+		to_chat(user, SPAN_NOTICE("You [unlocked ? "screw" : "unscrew"] the battery panel."))
+		unlocked = !unlocked
+
+	else if(W.is_crowbar() && unlocked)
+		to_chat(user, open ? \
+			"<span class='notice'>You crowbar the battery panel in place.</span>" : \
+			"<span class='notice'>You remove the battery panel.</span>" \
+		)
+		open = !open
+	
+	else if(W.is_wrench())
+		if(on)
+			to_chat(user, "<span class='notice'>\The [src] is active. Turn it off before trying to move it!</span>")
+			return FALSE
+		default_unfasten_wrench(user, W, 2 SECONDS)
+
+	else if(istype(W, /obj/item/weapon/cell) && open)
+		if(istype(cell))
+			to_chat(user, "<span class='notice'>There is a power cell already installed.</span>")
+			return FALSE
+		user.drop_from_inventory(W, src)
+		to_chat(user, "<span class='notice'>You insert the power cell.</span>")
+
+	else
+		. = ..()
+
+	RefreshParts()
+	update_icon()
+
+/obj/machinery/pump/proc/pump_reagents()
```

Comment:
This should really be a proc on the turf.

Reply:
Moved it onto turf. I would nominally like to make it an assoc list on the turf of {reagents : volume mult}, but because there's extra logic involved on water, that feels like an even clunkier solution to split the definition between variables and external environment parameters.
Also, loc.name = "Sea" is pretty bad, but this isn't much better. Put up https://github.com/PolarisSS13/Polaris/issues/8324 to address repathing some of this stuff because that's gonna be pretty big map changes, and probably pretty messy.